mgmtsystem_kpi/models/mgmtsystem_kpi.py

```diff
@@ -0,0 +1,208 @@
+    ), 'KPI Computation Type')
+
+    dbsource_id = fields.Many2one(
+        'base.external.dbsource',
+        'External DB Source',
+    )
+    kpi_code = fields.Text(
+        'KPI Code',
+        help=("SQL code must return the result as 'value' "
+              "(i.e. 'SELECT 5 AS value')."),
+    )
+    history_ids = fields.One2many(
+        'mgmtsystem.kpi.history',
+        'kpi_id',
+        'History',
+    )
+    active = fields.Boolean(
+        'Active',
+        help=("Only active KPIs will be updated by the scheduler based on"
+              " the periodicity configuration."), default=True
+    )
+    company_id = fields.Many2one(
+        'res.company', 'Company',
+        default=lambda self: self.env.user.company_id.id)
+
+    def _compute_display_last_kpi_value(self):
+        result = {}
+        for obj in self.browse(self.ids):
+            if obj.history_ids:
+                result[obj.id] = obj.history_ids[0].value
+            else:
+                result[obj.id] = 0
+
+        return result
+
+    @api.multi
+    def compute_kpi_value(self):
+        for obj in self:
+            kpi_value = 0
+            if obj.kpi_type == 'local' and is_select_query(obj.kpi_code):
+                self.env.cr.execute(obj.kpi_code)
+                dic = self.env.cr.dictfetchall()
+                if is_one_value(dic):
+                    kpi_value = dic[0]['value']
+            elif (obj.kpi_type == 'external' and obj.dbsource_id.id and
+                  is_select_query(obj.kpi_code)):
+                dbsrc_obj = obj.dbsource_id
+                res = dbsrc_obj.execute(obj.kpi_code)
+                if is_one_value(res):
+                    kpi_value = res[0]['value']
+            elif obj.kpi_type == 'python':
+                kpi_value = eval(obj.kpi_code)
+
+            threshold_obj = obj.threshold_id
+            values = {
+                'kpi_id': obj.id,
+                'value': kpi_value,
+                'color': threshold_obj.get_color(kpi_value),
+            }
+
+            history_obj = self.env['mgmtsystem.kpi.history']
+            history_id = history_obj.create(values)
+            obj.history_ids.append(history_id)
+
+        return True
+
+    def update_next_execution_date(self):
+        for obj in self:
+            if obj.periodicity_uom == 'hour':
+                delta = timedelta(hours=obj.periodicity)
+            elif obj.periodicity_uom == 'day':
+                delta = timedelta(days=obj.periodicity)
+            elif obj.periodicity_uom == 'week':
+                delta = timedelta(weeks=obj.periodicity)
+            elif obj.periodicity_uom == 'month':
+                delta = timedelta(months=obj.periodicity)
+            else:
+                delta = timedelta()
+            new_date = datetime.now() + delta
+
+            values = {
+                'next_execution_date': new_date.strftime(DATETIME_FORMAT),
+            }
+
+            obj.write(values)
```

Comment:
To write on a single field, it simple to use a For a simpler to use a direct assignment on `obj`.